src/MaxText/layers/moe.py

```diff
@@ -900,9 +905,15 @@ def gmm(inputs, kernel, group_sizes, expert_assignments):
 
     # w0, w1, wo needs to be un sharded on fsdp / fsdp_transpose axis, so use
     # mlp_no_fsdp axis
-    w0_pspec = nn.logical_to_mesh_axes(("exp", "embed_tensor_transpose", "mlp_no_fsdp"))
-    w1_pspec = nn.logical_to_mesh_axes(("exp", "embed_tensor_transpose", "mlp_no_fsdp"))
-    wo_pspec = nn.logical_to_mesh_axes(("exp", "mlp_no_fsdp", "embed_tensor_transpose"))
+    # special sharding for dsv3 to remove overhead between gmm/AG
+    if self.config.num_experts == 256:
```

Comment:
Similar comment with Matt. For instance, we could have a flag similar to `expert_shard_attention_option`, and we could have `expert_shard_mlp_option` or similar. 

This condition `self.config.num_experts == 256` may raise up question, how about 128 or 512 experts?

Reply:
I added shard_by_num_of_expert to only enable it when it expert_dim dividable by fdsp_shard_dim